shelley/chain-and-ledger/executable-spec/src/PoolRank.hs

```diff
@@ -0,0 +1,95 @@
+module PoolRank
+  ( desirability
+  , rankedPools
+  ) where
+
+import           BaseTypes (intervalValue)
+import           Coin (Coin (..))
+import           Data.Map.Strict (Map)
+import qualified Data.Map.Strict as Map
+import           Data.Ratio ((%))
+import qualified Data.Set as Set
+import           EpochBoundary (BlocksMade (..), SnapShots (..), Stake (..), poolStake)
+import           Keys (KeyHash)
+import           LedgerState (AccountState (..), EpochState (..), NewEpochState (..))
+import           Numeric.Natural (Natural)
+import           PParams (PParams (..))
+import           Slot (EpochSize)
+import           TxData (PoolParams (..))
+
+
+-- | Desirability calculation for non-myopic utily,
+-- corresponding to f^~ in section 5.6.1 of
+-- "Design Specification for Delegation and Incentives in Cardano"
+desirability
+  :: PParams
+  -> Coin
+  -> Natural
+  -> Natural
+  -> PoolParams crypto
+  -> Stake crypto
+  -> Coin
+  -> Rational
+desirability pp r blocksN blocksTotal pool (Stake stake) (Coin total) =
+  if fTilde <= cost
+    then 0
+    else (fTilde - cost)*(1 -margin)
+  where
+    fTilde      = fTildeNumer / fTildeDenom
+    fTildeNumer = pBar * fromIntegral r * (z0 + min s z0 * a0)
+    fTildeDenom = 1 + a0
+
+    cost   = fromIntegral $ _poolCost pool
+    margin = intervalValue $ _poolMargin pool
+    tot = max 1 (fromIntegral total)
+
+    Coin pledge = _poolPledge pool
+    s = fromIntegral pledge % tot
+    a0 = _a0 pp
+    z0 = 1 % max 1 (fromIntegral (_nOpt pp))
+
+    pBar = if sigma == 0
+             then 0
+           else if intervalValue (_d pp) < 0.8
+             then beta / sigma
+           else 1
+    Coin pstake = sum stake
+    sigma = fromIntegral pstake % tot
+    beta = fromIntegral blocksN / fromIntegral (max 1 blocksTotal)
+
+-- | Computes the top ranked stake pools
```

Comment:
Doesn't this compute a mapping from pool hash keys to desirability?
Maybe it would make sense to return a list of pool hash keys in the ranked order here?

Reply:
yes, you are right, it isn't returning the rank but the score. and we will need the rank to determine the non-myopic stake. thanks!